File: trie.go
package main

// Taken from https://medium.com/@itachisasuke/implementing-a-search-engine-in-golang-trie-data-structure-c45152ddda24

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// Node represent each character
type Node struct {
	//this is a single letter stored for example letter a,b,c,d,etc
	Char string
	//store all children  of a node
	//that is from a-z
	//a slice of Nodes(and each child will also have 26 children)
	Children [26]*Node
	IsWord   bool
}

// / NewNode this will be used to initialize a new node with 26 children
// /each child should first be initialized to nil
func NewNode(char string) *Node {
	node := &Node{Char: char, IsWord: false}
	for i := 0; i < 26; i++ {
		node.Children[i] = nil
	}

	return node
}

// Trie  is our actual tree that will hold all of our nodes
// the Root node will be nil
type Trie struct {
	RootNode *Node
}

// NewTrie Creates a new trie with a root('constructor')
func NewTrie() *Trie {
	//we will not use this node so
	//it can be anything
	root := NewNode("\000")
	return &Trie{RootNode: root}
}

// / Insert inserts a word to the trie
func (t *Trie) Insert(word string) error {
	///this will keep track of our current node
	///when transversing our tree
	///it should always start at the top of our tree
	///i.e our root
	current := t.RootNode
	///remove all spaces from the word
	///and convert it to lowercase
	strippedWord := strings.ToLower(strings.ReplaceAll(word, " ", ""))
	for i := 0; i < len(strippedWord); i++ {
		//from the ascii table a represent decimal number 97
		//from the ascii table b represent decimal number 98
		//from the ascii table c represent decimal number 99
		/// and so on so basically if we were to say  98-97=1 which means the index of b is 1 and for  c is 99-97
		///that what is happening below (we are taking the decimal representation of a character and subtracting decimal representation of a)
		index := strippedWord[i] - 'a'
		///check if current already has a node created at our current node
		//if not create the node
		if current.Children[index] == nil {
			current.Children[index] = NewNode(string(strippedWord[i]))
		}
		current = current.Children[index]
	}
	current.IsWord = true
	return nil
}

// SearchWord will return false if a word we
// are searching for is not in the trie
// and true otherwise
func (t *Trie) SearchWord(word string) bool {
	strippedWord := strings.ToLower(strings.ReplaceAll(word, " ", ""))
	current := t.RootNode
	for i := 0; i < len(strippedWord); i++ {
		index := strippedWord[i] - 'a'
		//we have encountered null in the path we were transversing meaning this is the last node
		///that means this word is not indexed(present) in this trie
		if current == nil || current.Children[index] == nil {
			return false
		}
		current = current.Children[index]
	}
	return current.IsWord
}

// isLeaf returns a boolean representing if this node does not have children.
func (t *Trie) isLeaf(word string) bool {
	strippedWord := strings.ToLower(strings.ReplaceAll(word, " ", ""))
	current := t.RootNode
	for i := 0; i < len(strippedWord); i++ {
		index := strippedWord[i] - 'a'
		if current == nil || current.Children[index] == nil {
			return true
		}
		current = current.Children[index]
	}
	for i := 0; i < 26; i++ {
		if current.Children[i] != nil {
			return false
		}
	}
	return true
}

func (t *Trie) populate() {
	// Open the file
	file, err := os.Open("CollinsScrabbleDictionary2019.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	// Create a scanner to read the file line by line
	scanner := bufio.NewScanner(file)

	// Iterate through each line and call the Insert function
	for scanner.Scan() {
		word := scanner.Text()
		t.Insert(word)
	}

	// Check for any errors during scanning
	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
	}
}

File: trie.cpp
// C++ implementation of search and insert
// operations on Trie
#include "trie.h"
using namespace std;

const int ALPHABET_SIZE = 26;

// trie node
struct TrieNode
{
	struct TrieNode *children[ALPHABET_SIZE];

	// isEndOfWord is true if the node represents
	// end of a word
	bool isEndOfWord;
};

// Returns new trie node (initialized to NULLs)
struct TrieNode *getNode(void)
{
	struct TrieNode *pNode = new TrieNode;

	pNode->isEndOfWord = false;

	for (int i = 0; i < ALPHABET_SIZE; i++)
		pNode->children[i] = NULL;

	return pNode;
}

// If not present, inserts key into trie
// If the key is prefix of trie node, just
// marks leaf node
void insert(struct TrieNode *root, string key)
{
	struct TrieNode *pCrawl = root;
	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->children[index])
			pCrawl->children[index] = getNode();

		pCrawl = pCrawl->children[index];
	}
	// mark last node as leaf
	pCrawl->isEndOfWord = true;
}

// Returns true if key presents in trie, else
// false
bool search(struct TrieNode *root, string key)
{
	struct TrieNode *pCrawl = root;

	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->children[index])
			return false;

		pCrawl = pCrawl->children[index];
	}

	return (pCrawl->isEndOfWord);
}

// Returns true if there's no children and this word ends on a leaf node
bool isLeaf(struct TrieNode *root, string key)
{
	struct TrieNode *pCrawl = root;

	for (int i = 0; i < key.length(); i++)
	{
		int index = key[i] - 'a';
		if (!pCrawl->children[index])
			return true;

		pCrawl = pCrawl->children[index];
	}

	for (int i = 0; i < ALPHABET_SIZE; i++) {
        if (pCrawl->children[i]) {
            return false;
        }
    }
    return true;
}

struct TrieNode* create_dictionary()
{
    struct TrieNode *root = getNode();
	ifstream input_file("CollinsScrabbleDictionary2019.txt");
    assert(input_file.is_open());
    string line;
    getline(input_file, line);
    while (!input_file.rdstate()) {
        transform(line.begin(), line.end(), line.begin(), [](unsigned char c){ return tolower(c); });
        insert(root, line);
        getline(input_file, line);
    }
    transform(line.begin(), line.end(), line.begin(), [](unsigned char c){ return tolower(c); });
    insert(root, line);
    input_file.close();

    return root;
}